test/transport_end2end_test.go

```diff
@@ -0,0 +1,198 @@
+// The closeCh in conn can be used to detect when conn is closed.
+type transportRestartCheckCreds struct {
+	mu          sync.Mutex
+	connections []connWrapperWithCloseCh
+}
+
+func (c *transportRestartCheckCreds) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	return rawConn, nil, nil
+}
+func (c *transportRestartCheckCreds) ClientHandshake(ctx context.Context, authority string, rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	conn := &connWrapperWithCloseCh{Conn: rawConn, closeCh: make(chan interface{}, 1)}
+	c.connections = append(c.connections, *conn)
+	commonAuthInfo := credentials.CommonAuthInfo{SecurityLevel: credentials.NoSecurity}
+	authInfo := &authInfoWithConn{commonAuthInfo, conn}
+	return conn, authInfo, nil
+}
+func (c *transportRestartCheckCreds) Info() credentials.ProtocolInfo {
+	return credentials.ProtocolInfo{}
+}
+func (c *transportRestartCheckCreds) Clone() credentials.TransportCredentials {
+	return &transportRestartCheckCreds{}
+}
+func (c *transportRestartCheckCreds) OverrideServerName(s string) error {
+	return nil
+}
+
+// Tests that the client transport drains and restarts when next stream ID exceeds
+// MaxStreamID. This test also verifies that subsequent RPCs use a new client
+// transport and the old transport is closed.
+func (s) TestClientTransportRestartsAfterStreamIDExhausted(t *testing.T) {
+	// Set the transport's MaxStreamID to 5 to cause connection to drain after 2 RPCs.
+	originalMaxStreamID := transport.MaxStreamID
+	transport.MaxStreamID = 5
+	defer func() {
+		transport.MaxStreamID = originalMaxStreamID
+	}()
+
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			for i := 0; i < 2; i++ {
+				if _, err := stream.Recv(); err != nil {
+					return status.Errorf(codes.Internal, "unexpected error receiving: %v", err)
+				}
+				if err := stream.Send(&testpb.StreamingOutputCallResponse{}); err != nil {
+					return status.Errorf(codes.Internal, "unexpected error sending: %v", err)
+				}
+			}
+			if recv, err := stream.Recv(); err != io.EOF {
+				return status.Errorf(codes.Internal, "Recv = %v, %v; want _, io.EOF", recv, err)
+			}
+			return nil
+		},
+	}
+
+	creds := &transportRestartCheckCreds{}
+	if err := ss.Start(nil, grpc.WithTransportCredentials(creds)); err != nil {
+		t.Fatalf("Starting stubServer: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	var streams []testpb.TestService_FullDuplexCallClient
+
+	// expectedNumConns when each stream is created.
+	expectedNumConns := []int{1, 1, 2}
+
+	// Set up 3 streams and call sendAndReceive() once on each.
+	for i := 0; i < 3; i++ {
+		s, err := ss.Client.FullDuplexCall(ctx)
+		if err != nil {
+			t.Fatalf("Creating FullDuplex stream: %v", err)
+		}
+
+		streams = append(streams, s)
```

Comment:
Is the connection guaranteed to be established even without this?

I think so, as gRPC won't return from `NewStream` unless there's at least a _connection_ for the RPC.  But maybe not, because it could be that we see the first connection as okay and return a stream that uses it

So _maybe_ we don't need this?  Or if we do need it, please add a comment explaining why.